presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/json/JsonFileBasedInMemoryFunctionNameSpaceManager.java

```diff
@@ -0,0 +1,216 @@
+import com.facebook.presto.spi.function.SqlInvokedFunction;
+import com.google.common.collect.ImmutableList;
+
+import javax.inject.Inject;
+
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.facebook.presto.common.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.plugin.base.JsonUtils.parseJson;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_USER_ERROR;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_ARGUMENTS;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.facebook.presto.spi.function.FunctionVersion.notVersioned;
+import static com.facebook.presto.spi.function.RoutineCharacteristics.Language.CPP;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.MoreCollectors.onlyElement;
+import static java.lang.Long.parseLong;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class JsonFileBasedInMemoryFunctionNameSpaceManager
+        extends AbstractSqlInvokedFunctionNamespaceManager
+{
+    private final Map<SqlFunctionId, SqlInvokedFunction> latestFunctions = new ConcurrentHashMap<>();
+    private final Map<QualifiedObjectName, UserDefinedType> userDefinedTypes = new ConcurrentHashMap<>();
+    private final JsonFileBasedFunctionNameSpaceManagerConfig managerConfig;
+
+    @Inject
+    public JsonFileBasedInMemoryFunctionNameSpaceManager(
+            @ServingCatalog String catalogName,
+            SqlFunctionExecutors sqlFunctionExecutors,
+            SqlInvokedFunctionNamespaceManagerConfig config,
+            JsonFileBasedFunctionNameSpaceManagerConfig managerConfig)
+    {
+        super(catalogName, sqlFunctionExecutors, config);
+        this.managerConfig = requireNonNull(managerConfig, "managerConfig is null");
+        bootStrapNameSpaceFromFile();
+    }
+
+    private static SqlInvokedFunction copyFunction(SqlInvokedFunction function)
+    {
+        return new SqlInvokedFunction(
+                function.getSignature().getName(),
+                function.getParameters(),
+                function.getSignature().getReturnType(),
+                function.getDescription(),
+                function.getRoutineCharacteristics(),
+                function.getBody(),
+                function.getVersion());
+    }
+
+    private void bootStrapNameSpaceFromFile()
+    {
+        if (managerConfig.getFunctionDefinitionFile().isEmpty()) {
+            return;
+        }
+        JsonBasedUdfFunctionSignatureMap jsonBasedUdfFunctionSignatureMap = parseJson(Paths.get(managerConfig.getFunctionDefinitionFile()), JsonBasedUdfFunctionSignatureMap.class);
```

Comment:
what if the file is not found or the json is invalid?

Reply:
It will throw exception. 

This function is called when construct the `JsonFileBasedFunctionNamespaceManager`, which happens here https://github.com/prestodb/presto/blob/19a9bb8a516ced26c8528f50b5d3f4cfa9d245a1/presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java#L164-L179

Do I need to add logic to capture exception and avoid tearing down the program here?

Reply:
Update: just add code to catch the exception, and log the information in logger.